arcs/src/components/spatial_entity.rs

```diff
@@ -0,0 +1,145 @@
+use crate::{
+    components::BoundingBox,
+    Vector,
+    primitives::{Arc},
+    algorithms::{Bounded},
+};
+use specs::{Entity, world::Index};
+use aabb_quadtree::{QuadTree, Spatial, ItemId};
+use quadtree_euclid::{TypedRect, TypedPoint2D, TypedSize2D};
+use std::collections::HashMap;
+
+pub(crate) type SpatialTree = QuadTree<SpatialEntity, f64, [(ItemId, TypedRect<f32, f64>); 0]>;
+
+/// A intermediate struct that maps an [`Entity`] to its [`BoundingBox`]
+/// 
+/// This is used to populate an efficient spatial lookup structure like a `QuadTree`
+#[derive(Debug)]
+pub struct SpatialEntity {
+    pub bounds: BoundingBox,
+    pub entity: Entity
+}
+
+impl Spatial<f64> for SpatialEntity {
+    fn aabb(&self) -> TypedRect<f32, f64> {
+        let bb = self.bounds;
+        TypedRect::<f32, f64>::new(
+            // TypedRects have their origin at the bottom left corner (this is undocumented!)
+            TypedPoint2D::new(bb.bottom_left().x as f32, bb.bottom_left().y as f32),
+            TypedSize2D::new(bb.width() as f32, bb.height() as f32))
+    }
+}
+
+impl SpatialEntity {
+    pub fn new(bounds: BoundingBox, entity: Entity) -> SpatialEntity {
+        SpatialEntity {
+            bounds,
+            entity
+        }
+    }
+}
+
+/// A global [`Resource`] for looking up which [`Entity`]s inhabit
+/// a given spatial point or region
+#[derive(Debug)]
+pub struct Space {
+    quadtree: SpatialTree,
+    ids: HashMap<Index, ItemId>
+}
+
+impl Default for Space {
+    fn default() -> Self {
+        Space {
+            quadtree: Self::default_tree(),
+            ids: HashMap::new()
+        }
+    }
+}
+
+impl Space {
+    // FIXME: Hard-code is bad-bad
+    const WORLD_RADIUS: f64 = 1_000_000.0;
+
+    // FIXME: We need to supply this in *DrawingUnits*
+    // as 1[meter] f.e. becomes meaningless when zoomed far in/out
+    const QUERY_POINT_RADIUS: f64 = 1.0;
+
+    fn default_tree() -> SpatialTree{
+        // Initialize quadtree
+        let size = BoundingBox::new(
+            Vector::new(-Self::WORLD_RADIUS, -Self::WORLD_RADIUS),
+            Vector::new(Self::WORLD_RADIUS, Self::WORLD_RADIUS)
+            ).aabb();
+        let quadtree: SpatialTree = QuadTree::new(
+            size,
+            true,
+            4,
+            16,
+            8,
+            4
+        );
+
+        quadtree
+    }
+
+    pub fn insert(&mut self, spatial: SpatialEntity) {
+        let entity_id = spatial.entity.id();
+        if let Some(id) = self.quadtree.insert(spatial) {
+            self.ids.insert(entity_id, id);
+        }
+    }
+
+    pub fn modify(&mut self, spatial: SpatialEntity) {
+        let entity_id = spatial.entity.id();
+        if self.ids.contains_key(&entity_id) {
```

Comment:
The idea behind this is that we can't insert the same entity twice? If it already exists we modify it instead?

If we keep two different methods I would agree on throwing an error when trying to modify a `SpatialEntity` which is not already present.

It would be less *explicit* but probably more *usage-friendly* to merge those two methods.

Reply:
I merged the two methods